File: EzTG.php
<?php
set_time_limit(0);
class EzTGException extends Exception {
}
class EzTG {
  private $settings;
  private $offset;
  private $json_payload;
  public function __construct($settings, $base = false) {
    $this->settings = array_merge(array(
      'endpoint' => 'https://api.telegram.org',
      'token' => '1234:abcd',
      'callback' => function($update, $EzTG) {
        echo 'no callback' . PHP_EOL;
      },
      'objects' => true,
      'throw_telegram_errors' => true,
      'magic_json_payload' => true
    ), $settings);
    if ($base !== false)
      return true;
    if (!is_callable($this->settings['callback']))
      $this->error('Invalid callback.', true);
    if (php_sapi_name() === 'cli') {
      $this->settings['magic_json_payload'] = false;
      $this->offset = -1;
      $this->get_updates();
    } else {
      if ($this->settings['magic_json_payload'] === true) {
        ob_start();
        $this->json_payload = false;
        register_shutdown_function(array($this, 'send_json_payload'));
      }
      $this->processUpdate(json_decode(file_get_contents('php://input')));
    }
  }
  private function get_updates() {
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $this->settings['endpoint'] . '/bot' . $this->settings['token'] . '/getUpdates');
    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    while (true) {
      curl_setopt($ch, CURLOPT_POSTFIELDS, 'offset=' . $this->offset . '&timeout=10');
      if ($this->settings['objects'] === true) {
        $result = json_decode(curl_exec($ch));
        if (isset($result->ok) and $result->ok === false)
          $this->error($result->description, false);
        elseif (isset($result->result)) {
          foreach ($result->result as $update) {
            if (isset($update->update_id))
              $this->offset = $update->update_id + 1;
            $this->processUpdate($update);
          }
        }
      } else {
        $result = json_decode(curl_exec($ch), true);
        if (isset($result['ok']) and $result['ok'] === false)
          $this->error($result['description'], false);
        elseif (isset($result['result'])) {
          foreach ($result['result'] as $update) {
            if (isset($update['update_id']))
              $this->offset = $update['update_id'] + 1;
            $this->processUpdate($update);
          }
        }
      }
    }
  }
  public function processUpdate($update) {
    $this->settings['callback']($update, $this);
  }
  protected function error($e, $throw = 'default') {
    if ($throw === 'default')
      $throw = $this->settings['throw_telegram_errors'];
    print_r($throw);
    if ($throw === true)
      throw new EzTGException($e);
    else {
      echo 'Telegram error: ' . $e . PHP_EOL;
      return array(
        'ok' => false,
        'description' => $e
      );
    }
  }
  public function newKeyboard($type = 'keyboard', $rkm = array('resize_keyboard' => true, 'keyboard' => array())) {
    return new EzTGKeyboard($type, $rkm);
  }
  public function __call($name, $arguments) {
    if (!isset($arguments[0]))
      $arguments[0] = array();
    if (!isset($arguments[1]))
      $arguments[1] = true;
    if ($this->settings['magic_json_payload'] === true and $arguments[1] === true) {
      if ($this->json_payload === false) {
        $arguments[0]['method'] = $name;
        $this->json_payload = $arguments[0];
        return 'json_payloaded'; //xd
      } elseif(is_array($this->json_payload)) {
        $old_payload = $this->json_payload;
        $arguments[0]['method'] = $name;
        $this->json_payload = $arguments[0];
        $name = $old_payload['method'];
        $arguments[0] = $old_payload;
        unset($arguments[0]['method']);
        unset($old_payload);
      }
    }
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $this->settings['endpoint'] . '/bot' . $this->settings['token'] . '/' . urlencode($name));
    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($arguments[0]));
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    if ($this->settings['objects'] === true)
      $result = json_decode(curl_exec($ch));
    else
      $result = json_decode(curl_exec($ch), true);
    curl_close($ch);
    if ($this->settings['objects'] === true) {
      if (isset($result->ok) and $result->ok === false)
        return $this->error($result->description);
      if (isset($result->result))
        return $result->result;
    } else {
      if (isset($result['ok']) and $result['ok'] === false)
        return $this->error($result['description']);
      if (isset($result['result']))
        return $result['result'];
    }
    return $this->error('Unknown error', false);
  }
  public function send_json_payload() {
    if (is_array($this->json_payload)) {
      ob_end_clean();
      echo json_encode($this->json_payload);
      header('Content-Type: application/json');
      ob_end_flush();
      return true;
    }
  }
}
class EzTGKeyboard {
  public function __construct($type = 'keyboard', $rkm = array('resize_keyboard' => true, 'keyboard' => array())) {
    $this->line = 0;
    $this->type = $type;
    if ($type === 'inline')
      $this->keyboard = array(
        'inline_keyboard' => array()
      );
    else
      $this->keyboard = $rkm;
    return $this;
  }
  public function add($text, $callback_data = NULL, $type = 'auto') {
    if ($this->type === 'inline') {
      if ($callback_data === NULL)
        $callback_data = trim($text);
      if (!isset($this->keyboard['inline_keyboard'][$this->line]))
        $this->keyboard['inline_keyboard'][$this->line] = array();
      if ($type === 'auto')
        if (filter_var($callback_data, FILTER_VALIDATE_URL))
          $type = 'url';
        else
          $type = 'callback_data';
      array_push($this->keyboard['inline_keyboard'][$this->line], array(
        'text' => $text,
        $type => $callback_data
      ));
    } else {
      if (!isset($this->keyboard['keyboard'][$this->line]))
        $this->keyboard['keyboard'][$this->line] = array();
      array_push($this->keyboard['keyboard'][$this->line], $text);
    }
    return $this;
  }
  public function newline() {
    $this->line++;
    return $this;
  }
  public function done() {
    if ($this->type === 'remove')
      return '{"remove_keyboard": true}';
    else
      return json_encode($this->keyboard);
  }
}

File: Python/Standard/EzTG.py
import typing
import requests
import urllib.parse
import warnings
import json


class APIException(Exception):
    pass


class EzTG:
    def __init__(self, token: str, callback: typing.Callable, base=False, throw_telegram_errors=True, endpoint='https://api.telegram.org', offset: int = -1, timeout: int = 10):
        self.token = token
        self.callback = callback
        self.endpoint = endpoint
        self.offset = offset
        self.timeout = timeout
        self.throw_telegram_errors = throw_telegram_errors
        if base == False:  # base serve x non fare getupdates e wuindi se non serve pigliare mex
            self._getUpdates()

    def _getUpdates(self):
        updates = {}
        while True:
            try:
                updates = self._telegramRequest(
                    'getUpdates', {'offset': self.offset, 'timeout': self.timeout})
            except Exception as e:
                warnings.warn(str(e))
            for update in updates:
                self.offset = update['update_id'] + 1
                self.processUpdate(update)

    def processUpdate(self, update):
        return self.callback(self, update)

    def _telegramRequest(self, method, params={}):
        r = requests.post(self.endpoint + '/bot' + self.token +
                          '/' + urllib.parse.quote(method), data=params)
        response = r.json()
        if response['ok'] == True and 'result' in response:
            return response['result']
        else:
            if response['ok'] == False and 'description' in response:
                if self.throw_telegram_errors:
                    raise APIException(response['description'])
                else:
                    warnings.warn(response['description'])
            else:
                if self.throw_telegram_errors:
                    raise APIException('Unknown error')
                else:
                    warnings.warn('Unknown error')
            return response

    def __getattr__(self, name):
        def function(**kwargs):
            return self._telegramRequest(name, kwargs)
        return function


class Keyboard:
    def __init__(self, type='keyboard', rkm={'resize_keyboard': True, 'keyboard': []}):
        self.line = []
        self.type = type
        if type == 'inline':
            rkm = {'inline_keyboard': []}
        self.keyboard = rkm

    def add(self, text: str, callback_data=None, type='auto'):
        if self.type == 'inline':
            if callback_data == None:
                callback_data = text
            if type == 'auto':
                if callback_data.startswith('http://') or callback_data.startswith('https://'):
                    type = 'url'
                else:
                    type = 'callback_data'
            self.line.append(
                {'text': text, type: callback_data})
        else:
            self.line.append(text)
        return self

    def newLine(self):
        if self.type == 'inline':
            self.keyboard['inline_keyboard'].append(self.line)
        else:
            self.keyboard['keyboard'].append(self.line)
        self.line = []
        return self

    def done(self):
        self.newLine()
        if self.type == 'remove':
            return '{"remove_keyboard": true}'
        else:
            return json.dumps(self.keyboard)

    def __str__(self):
        return self.done()
